### ISSUE TITLE
Add option to export animation to SVG format

### ISSUE THREAD

**Issue by Emasoft** [opentoonz/opentoonz_docs#148]
For animations made using the vector layer, it would be convenient to add an option to export them in SVG format. It has the following advantages:
- It is an open standard for vector graphics
- OpenToonz can already import/export vector layers in SVG format
- You can edit the SVG vector layer in other software tools (Inkscape, Open Office, Illustrator, Affinity Designer, Autodesk Graphics, CACANi, etc.)
- You can easily create a playable animation from the SVG frames. Just paste each svg frame as a group in a svg file, and add this animate tag inside each group:

``` xml
<g id="frame60"  style="display:inline">
<animate
       attributeName="display"
       fill="freeze"
       repeatCount="indefinite"
       dur="1.3s"
       keyTimes="0;0.961;0.992;1"
       values="none;inline;none;none" />
...
<g id="frame60content" style="display:inline">
 (...paste the OpenToonz svg frame here...)
</g>
...
</g>
```
- You can preview the vector animation on every browser.

I've made an animation test, joining the exported SVG in Inkscape (I just pasted each frame exported from OpenToonz in SVG as a new inkscape layer, and saved. But a script can do this process much quicker). 

You can playback and download the resulting svg animation from here: 

http://imgh.us/navetta_test.svg

**Comment by erayzesen** [comment]
Hi all. This topic is amazing. The animated svg export feature can be a bomb feature for OT. What happened about this request?

**Comment by RodneyBaker** [comment]
Hi @erayzesen and thanks for the interest!

There have been some minor improvements to SVG export; specificially to address output filenaming and to resolve an issue with Opentoonz getting confused with the PLI extension as opposed to the SVG extension.  These improvements started in the Tahoma2D fork and were then ported over to Opentoonz.

It probably should be restated that Opentoonz does not actually support SVG but rather converts to and from SVG.   This conversion then allows the user to work with the native Opentoonz PLI vector format.

@Emasoft's contribution to this works with the SVG files themselves; using python to take a series of SVG files and map them into one single SVG file formatted for frame animation.  This is done independently of Opentoonz.

That's basically where we are now with this request.

The most likely area where Opentoonz would see additional SVG enhancements is probably the OCA (Open Cel Animation) export option but currently there is no one known to be working on that at the moment.

**Comment by Emasoft** [comment]
I will start working again on the FBF SVG encoder tomorrow. I was busy with work and I didn't do much progress on it lately, but I should have free time to complete the script in the next few weeks. I hope that the FBFSVG format will become the standard export format of Opentoonz for vector animations. While the AI revolution is acting as a disruptor for all raster animations apps, in the case of vector animations the AI is not going to disrupt, because the current generations of AI is (ironically) very bad at math. AI can at best be used as a post processing step when rasterizing the vector animation, but the core of any animation project will be vector based since it must be continuosly editable and each scene must be separable in parts that independent animators can work on at the same time without worring about interfering with the others. Opentoonz should really make the fbf.svg file an integral part of the Opentoonz project file format. There is nothing more open than svg after all. I also already included support in the fbf script for the gradient meshes (the one feature that Adobe with a trick stopped to become part of the official SVG standard because it waa the only feature that Adobe Illustrator .ai format had more than SVG, and it would threaten its market; but that luckily Inkscape already supported), so there is really no limit to what SVG can encode in vector format.

**Comment by erayzesen** [comment]
@RodneyBaker  thanks for the quick reply. 

> The most likely area where Opentoonz would see additional SVG enhancements is probably the OCA (Open Cel Animation) export option but currently there is no one known to be working on that at the moment.

I had issues with importing my SVG. Since there were many reports on this, I was investigating the issues related to it without reopening. During SVG import trials, my only current problem was with scaling. It was strangely reducing the scale of the OT-imported SVG layers. So, it seemed like there were very few issues in translating from SVG to PLI, actually. If I'm not mistaken, work is ongoing on this matter


> @Emasoft's contribution to this works with the SVG files themselves; using python to take a series of SVG files and map them into one single SVG file formatted for frame animation. This is done independently of Opentoonz.

Still sounds great. Animated SVG is currently a miraculous format for export. It would be great if it had been integrated into this Opentoonz somehow.

@Emasoft  Your work is very nice, and if integrated into OpenToonz, it could be an amazing feature.

> I also believe Opentoonz should fully integrate the fbf.svg file format into its project files, and not only as an export format. SVG is an open standard, which makes it an ideal choice. Additionally, I've already added support for gradient meshes in the FBF script—a feature Adobe blocked from becoming part of the official SVG standard to protect its market share.

I agree with you. It would be great if vector layers in OpenToonz worked seamlessly with SVG. However, if I'm not mistaken, the processes at the vector level in OpenToonz operate with a significantly different approach than the standard SVG operations. Even if vector layers weren't fully integrated with SVG in OpenToonz, having a robust SVG import-export feature would be fantastic. For example, Blender's Grease Pencil side has more contradictions in its SVG approach. However, Blender has a very efficient 'import->SVG as Grease Pencil' feature when importing SVGs. Additionally, Blender can export each frame as SVG beautifully (though, of course, you can't export it as an animated SVG).

I wish I could have at least assisted with SVG import and found the time. It probably represents a significant obstacle for many people's pipelines. OpenToonz is a powerful tool for vector animation. The combination of the skeletal rig tool and the plastic tool in animation, with a sophisticated approach like frame-by-frame animation synthesis, has very few equivalents even as paid software. (I'm not even mentioning the robustness of the in-between interpolation algorithm.) If the issues related to SVG are overcome, and tools like alignment tools similar to those in Inkscape are added for vector editing, and features like rounded corners are added to the rectangle tool at the vector level, I can't think of any other software that compares.

**Comment by Emasoft** [comment]
@erayzesen I agree wholeheartedly and with bells on. Opentoonz (for 2D) and Blender (for 3D) are going to be the gold standards and the industry benchmarks in the field of open source computer animation. Especially if they add support for AI postprocessing in the form of SD AnimateDiff, automating the workflow with a simple UI to let the non tech savy artists to use it easily.

**Comment by RodneyBaker** [comment]
Moving this back to regular Opentoonz issues in light of recent fixes to SVG import and export.
While export of animated vector sequence to a single document frame by frame has not been implemented such things as the unfortunate offset of SVG imports and issues with SVG sequences as well as a host of other issues with SVG have been addressed.
